File: Problem_1 - 1.py
"""
Time Complexity : O(n^2). 
Space Complexity : O(n^2).
    
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not postorder:
            return None 
        
        root_val = postorder[-1]
        node = TreeNode(root_val)
        index = 0
        
        for i in range(len(inorder)):
            if root_val == inorder[i]:
                index = i
        
        inleft = inorder[:index]
        inright = inorder[index+1:]
        poleft = postorder[:index]
        poright = postorder[index:-1]
        
        node.left = self.buildTree(inleft,poleft)
        node.right = self.buildTree(inright,poright)
        
        return node 

File: Problem1_1.java
//Time Complexity: O(n^2)
//Space Complexity: O(n*h), where h is the height of the tree
//Code run successfully on LeetCode.

public class Problem1_1 {

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        
        if(postorder == null ||postorder.length == 0|| inorder == null|| inorder.length == 0)
            return null;
        
        int rootVal = postorder[postorder.length - 1];
        TreeNode root = new TreeNode(rootVal);
        int rootIdx = 0;
        
        for(int i =0; i< inorder.length; i++){
            if(inorder[i] == rootVal)
            {
                rootIdx = i;
                break;
            }
        }
        
        int[] inorderLeft = Arrays.copyOfRange(inorder, 0, rootIdx );
        int[] inorderRight = Arrays.copyOfRange(inorder, rootIdx +1 , inorder.length);
        int[] postorderLeft = Arrays.copyOfRange(postorder, 0 , rootIdx);
        int[] postorderRight = Arrays.copyOfRange(postorder, rootIdx  , postorder.length-1);
        
        root.left = buildTree(inorderLeft, postorderLeft);
        root.right = buildTree(inorderRight, postorderRight);
        
        return root;
    }
}
